packages/core/src/request-builder/header-builder/auth-headers.ts

```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
+import { MapType, errorWithCause, createLogger } from '@sap-cloud-sdk/util';
+import {
+  DestinationAuthToken,
+  Destination,
+  getOAuth2ClientCredentialsToken
+} from '../../scp-cf';
+import { isNullish, toHeaderObject } from './headers-util';
+
+const logger = createLogger({
+  package: 'core',
+  messageContext: 'auth-headers'
+});
+
+function toAuthorizationHeader(authorization: string): MapType<any> {
+  return toHeaderObject('authorization', authorization);
+}
+
+function headerFromTokens(authTokens?: DestinationAuthToken[] | null): string {
+  if (!authTokens || !authTokens.length) {
+    throw Error(
+      'AuthenticationType is "OAuth2SAMLBearerAssertion", but no AuthTokens could be fetched from the destination service!'
+    );
+  }
+
+  const usableTokens = authTokens.filter(
+    (token: DestinationAuthToken) => !token.error
+  );
+
+  if (!usableTokens.length) {
+    throw Error(
+      [
+        'The destination tried to provide authorization tokens but failed in all cases. This is most likely due to misconfiguration.',
+        'Original error messages:',
+        ...authTokens.map(token => token.error)
+      ].join('\n')
+    );
+  }
+  const authToken = usableTokens[0];
+  return `${authToken.type} ${authToken.value}`;
+}
+
+async function headerFromOAuth2ClientCredentialsDestination(
+  destination: Destination
+): Promise<string> {
+  const response = await getOAuth2ClientCredentialsToken(destination).catch(
+    error => {
+      throw errorWithCause(
+        'Request for "OAuth2ClientCredentials" authentication access token failed or denied.',
+        error
+      );
+    }
+  );
+  return `Bearer ${response.access_token}`;
+}
+
+function headerFromBasicAuthDestination(destination: Destination): string {
+  if (isNullish(destination.username) || isNullish(destination.password)) {
+    throw Error(
+      'AuthenticationType is "BasicAuthentication", but "username" and / or "password" are missing!'
+    );
+  }
+
+  return basicHeader(destination.username, destination.password);
+}
+
+export function basicHeader(username: string, password: string): string {
+  return `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`;
+}
+
+function headerForPrincipalPropagation(destination: Destination): MapType<any> {
```

Comment:
Minor: a few of the `headerForXXX` methods return a string and then a `toAuthHeader` or so is used to build the key-value header map. Others like this one return the map directly. Perhaps you can unify this. I know that I added this inconsistency partly.

Reply:
I thought of this and then forgot. Will do.